src/rsz/src/SizeDownMove.cc

```diff
@@ -172,168 +171,271 @@ bool SizeDownMove::doMove(const Path* drvr_path,
-                                          const DcalcAnalysisPt* dcalc_ap,
-                                          float fanout_slack)
+// This will downsize the gate to the smallest input capacitance that that
+// satisfies the given slack margin.
+LibertyCell* SizeDownMove::downSizeGate(const LibertyPort* drvr_port,
+                                        const LibertyPort* load_port,
+                                        const Pin* load_pin,
+                                        const DcalcAnalysisPt* dcalc_ap,
+                                        float slack_margin)
 {
   const int lib_ap = dcalc_ap->libertyIndex();
-  LibertyCell* fanout_cell = fanout_port->libertyCell();
-  const Instance* fanout_inst = network_->instance(fanout_pin);
+  LibertyCell* load_cell = load_port->libertyCell();
+  const char* load_port_name = load_port->name();
 
-  LibertyCellSeq swappable_cells = resizer_->getSwappableCells(fanout_cell);
+  LibertyCellSeq swappable_cells = BaseMove::getSwappableCells(load_cell);
   LibertyCell* best_cell = nullptr;
 
   if (swappable_cells.size() > 1) {
-    const char* fanout_port_name = fanout_port->name();
-
     // Sort from the smallest input capacitance to the smallest
     // breaking tie by the intrinsic delay
     sort(&swappable_cells,
          [=](const LibertyCell* cell1, const LibertyCell* cell2) {
            LibertyPort* port1
-               = cell1->findLibertyPort(fanout_port_name)->cornerPort(lib_ap);
+               = cell1->findLibertyPort(load_port_name)->cornerPort(lib_ap);
            const LibertyPort* port2
-               = cell2->findLibertyPort(fanout_port_name)->cornerPort(lib_ap);
+               = cell2->findLibertyPort(load_port_name)->cornerPort(lib_ap);
 
            const float cap1 = port1->capacitance();
            const float cap2 = port2->capacitance();
            const ArcDelay intrinsic1 = port1->intrinsicDelay(this);
            const ArcDelay intrinsic2 = port2->intrinsicDelay(this);
            return (std::tie(cap1, intrinsic2) < std::tie(cap2, intrinsic1));
          });
+  }
 
-    const float drvr_load_cap = graph_delay_calc_->loadCap(drvr_pin, dcalc_ap);
-    const float fanout_cap = fanout_port->cornerPort(lib_ap)->capacitance();
-
-    // Get fanouts based on Liberty since STA arcs are not present in DFFs
-    vector<const Pin*> fanouts = getFanouts(fanout_inst);
-
-    int max_fanout_id = 0;
-    float fanout_load_cap = -1.0;
-    // Find the output pin with max cap load and use that if more than one.
-    // We might want to use the most critical one instead?
-    for (int i = 0; i < fanouts.size(); i++) {
-      const float cur_fanout_load_cap
-          = graph_delay_calc_->loadCap(fanouts[i], dcalc_ap);
-      if (cur_fanout_load_cap > fanout_load_cap) {
-        fanout_load_cap = cur_fanout_load_cap;
-        max_fanout_id = i;
-      }
+  string swappable_names;
+  for (LibertyCell* swappable : swappable_cells) {
```

Comment:
Nit: ideally we would gate `swappable_names` creation on `logger->debugCheck(RSZ, "size_down", 3)`. It's fine to leave it as is

Reply:
Good call. I will do this.